File: go/identity_provider_config.go
// Package realmid — identity-provider config CRUD (admin surface).
//
// This is the realm-admin management client for federated identity
// providers (google / microsoft / facebook / apple). It is distinct
// from Realm.IdentityProviders, which is the public, read-only SPA
// discovery endpoint. The server requires platform_id on every call;
// the SDK injects it as the realm's own id so callers never pass it.
package realmid

import (
	ctxpkg "context"
	"errors"
	"net/url"
)

// IDPConfig is one identity-provider configuration row, as returned by
// the /identity-providers admin endpoints (wire = snake_case).
type IDPConfig struct {
	ID             string   `json:"id"`
	EntityType     string   `json:"entity_type"` // "realm" | "tenant"
	EntityID       string   `json:"entity_id"`
	Provider       string   `json:"provider"`    // "google" | "microsoft" | "facebook" | "apple"
	ClientType     string   `json:"client_type"` // "web" | "ios" | "android" | "desktop" | "other"
	ClientID       string   `json:"client_id"`
	AllowedOrigins []string `json:"allowed_origins"`
	Comments       string   `json:"comments"`
	// Config is the provider's PUBLIC config (never secrets) — e.g. the
	// Firebase web config (apiKey, authDomain, projectId, appId). It is
	// echoed verbatim on public discovery so a browser SDK can bootstrap
	// sign-in. Omitted when empty.
	Config    map[string]string `json:"config,omitempty"`
	Enabled   bool              `json:"enabled"`
	CreatedAt int64             `json:"created_at"`
	UpdatedAt int64             `json:"updated_at"`
}

// IDPConfigListPage is the list envelope for GET /identity-providers.
type IDPConfigListPage struct {
	Items []IDPConfig `json:"items"`
}

// IDPConfigListOpts are the optional filters for List.
type IDPConfigListOpts struct {
	// TenantID scopes the listing to a single tenant within the realm.
	// Empty lists realm-scope (and tenant) providers per server default.
	TenantID string
}

// IDPConfigCreate is the POST body. platform_id is injected by the SDK
// (= realm id) and must not be set here. When TenantID is set the
// provider is scoped to that tenant; otherwise it is realm-scoped.
//
// Validation (server-enforced, passed through): client_type=web REQUIRES
// a non-empty AllowedOrigins; non-web REQUIRES it absent/empty.
type IDPConfigCreate struct {
	TenantID       string
	Provider       string
	ClientType     string
	ClientID       string
	AllowedOrigins []string
	Comments       string
	// Config is the provider's PUBLIC config (never secrets) — e.g. the
	// Firebase web config (apiKey, authDomain, projectId, appId). Echoed
	// verbatim on public discovery. Optional; omit for plain OIDC.
	Config map[string]string
}

// IDPConfigPatch is the PATCH body. All fields are optional (pointer
// semantics — nil means "don't touch"). At least one must be set or the
// server returns empty_patch.
type IDPConfigPatch struct {
	Enabled        *bool
	ClientID       *string
	AllowedOrigins *[]string
	Comments       *string
	// Config, when non-nil, REPLACES the stored provider config map
	// wholesale (not merged). Publishable values only — never secrets.
	Config *map[string]string
}

// IDPConfigDeleteResult is the DELETE acknowledgment.
type IDPConfigDeleteResult struct {
	Status string `json:"status"`
}

// Typed errors mirrored from the server taxonomy. Use errors.Is to
// branch on these.
var (
	ErrIDPExists   = errors.New("realmid: identity provider already exists")
	ErrIDPNotFound = errors.New("realmid: identity provider not found")
)

// IdentityProviderConfigClient is realm.IdentityProviderConfig — the
// admin CRUD surface for federated identity providers. Authz uses the
// realm's platform token, like RolesClient.
type IdentityProviderConfigClient struct {
	realm *Realm
}

// List returns the identity-provider configs for this realm, optionally
// scoped to a tenant. A nil/absent items array normalizes to empty.
func (c *IdentityProviderConfigClient) List(ctx ctxpkg.Context, opts *IDPConfigListOpts) (*IDPConfigListPage, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	q := map[string]string{"platform_id": c.realm.realmID}
	if opts != nil && opts.TenantID != "" {
		q["tenant_id"] = opts.TenantID
	}
	var page IDPConfigListPage
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "GET",
		Path:   "/identity-providers",
		Bearer: tok,
		Query:  q,
	}, &page); err != nil {
		return nil, mapIDPErr(err)
	}
	if page.Items == nil {
		page.Items = []IDPConfig{}
	}
	return &page, nil
}

// Create registers a new identity-provider config. platform_id is
// injected as the realm id. Returns ErrIDPExists (409) when a config for
// the same (entity, provider, client_type) already exists.
func (c *IdentityProviderConfigClient) Create(ctx ctxpkg.Context, body IDPConfigCreate) (*IDPConfig, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	wire := map[string]any{
		"platform_id": c.realm.realmID,
		"provider":    body.Provider,
		"client_type": body.ClientType,
		"client_id":   body.ClientID,
	}
	if body.TenantID != "" {
		wire["tenant_id"] = body.TenantID
	}
	if len(body.AllowedOrigins) > 0 {
		wire["allowed_origins"] = body.AllowedOrigins
	}
	if body.Comments != "" {
		wire["comments"] = body.Comments
	}
	if len(body.Config) > 0 {
		wire["config"] = body.Config
	}
	var out IDPConfig
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "POST",
		Path:   "/identity-providers",
		Bearer: tok,
		Body:   wire,
	}, &out); err != nil {
		return nil, mapIDPErr(err)
	}
	return &out, nil
}

// Update patches an existing identity-provider config. Only the set
// (non-nil) patch fields are sent. At least one must be set or the
// server returns empty_patch. Returns ErrIDPNotFound (404) for an
// unknown id.
func (c *IdentityProviderConfigClient) Update(ctx ctxpkg.Context, id string, patch IDPConfigPatch) (*IDPConfig, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	wire := map[string]any{}
	if patch.Enabled != nil {
		wire["enabled"] = *patch.Enabled
	}
	if patch.ClientID != nil {
		wire["client_id"] = *patch.ClientID
	}
	if patch.AllowedOrigins != nil {
		wire["allowed_origins"] = *patch.AllowedOrigins
	}
	if patch.Comments != nil {
		wire["comments"] = *patch.Comments
	}
	if patch.Config != nil {
		wire["config"] = *patch.Config
	}
	var out IDPConfig
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "PATCH",
		Path:   "/identity-providers/" + url.PathEscape(id),
		Bearer: tok,
		Body:   wire,
	}, &out); err != nil {
		return nil, mapIDPErr(err)
	}
	return &out, nil
}

// Delete removes an identity-provider config. Returns ErrIDPNotFound
// (404) for an unknown id. The status defaults to "deleted" when the
// server omits it.
func (c *IdentityProviderConfigClient) Delete(ctx ctxpkg.Context, id string) (*IDPConfigDeleteResult, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var res IDPConfigDeleteResult
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "DELETE",
		Path:   "/identity-providers/" + url.PathEscape(id),
		Bearer: tok,
	}, &res); err != nil {
		return nil, mapIDPErr(err)
	}
	if res.Status == "" {
		res.Status = "deleted"
	}
	return &res, nil
}

// mapIDPErr translates server error envelopes to the typed sentinels.
// The provider-specific `code` strings aren't part of the canonical
// ErrorCode union, so we read them from the envelope siblings via
// detailCode (shared with mapRoleErr).
func mapIDPErr(err error) error {
	var re *RealmError
	if !errors.As(err, &re) {
		return err
	}
	switch detailCode(re) {
	case "provider_exists":
		return errors.Join(ErrIDPExists, re)
	case "provider_not_found":
		return errors.Join(ErrIDPNotFound, re)
	}
	if re.Code == ErrCodeNotFound {
		return errors.Join(ErrIDPNotFound, re)
	}
	if re.Code == ErrCodeConflict {
		return errors.Join(ErrIDPExists, re)
	}
	return re
}

File: ts/src/identity-provider-config.ts
/**
 * Identity-provider config CRUD — realm-admin resource client.
 *
 * This is the ADMIN management surface for social/OIDC provider configs
 * (Google, Microsoft, Facebook, Apple). It is DISTINCT from the public
 * IdP discovery surface; do not conflate the two.
 *
 * Authorization is the short-lived platform token, injected automatically
 * by the HttpClient (same model as `roles`). `platform_id` is required by
 * the server but is ALWAYS the realm's own id, so the SDK injects it
 * automatically — callers never pass it. An optional `tenantId` scopes a
 * provider to a single tenant within the realm.
 */

import type { HttpClient } from "./http.js";

export type IdpEntityType = "realm" | "tenant";
export type IdpProvider = "google" | "microsoft" | "facebook" | "apple";
export type IdpClientType = "web" | "ios" | "android" | "desktop" | "other";

export interface IdpConfig {
  id: string;
  entity_type: IdpEntityType;
  entity_id: string;
  provider: IdpProvider;
  client_type: IdpClientType;
  client_id: string;
  allowed_origins: string[];
  comments: string;
  /**
   * Provider-specific PUBLIC config (never secrets) — e.g. the Firebase
   * web config (`apiKey`, `authDomain`, `projectId`, `appId`). Echoed
   * verbatim on public discovery so a browser SDK can bootstrap sign-in.
   * Absent when empty.
   */
  config?: Record<string, string>;
  enabled: boolean;
  created_at: number;
  updated_at: number;
  [k: string]: unknown;
}

export interface IdpConfigListPage {
  items: IdpConfig[];
}

export interface IdpConfigListOpts {
  /** Scope the listing to a single tenant within the realm. */
  tenantId?: string;
}

export interface IdpConfigCreate {
  /** Optional tenant scope; omit for a realm-level provider. */
  tenantId?: string;
  provider: IdpProvider;
  clientType: IdpClientType;
  clientId: string;
  /**
   * Required (non-empty) when `clientType === "web"`; must be absent/empty
   * otherwise. The server enforces this; the SDK passes it through.
   */
  allowedOrigins?: string[];
  comments?: string;
  /**
   * Provider-specific PUBLIC config (never secrets) — e.g. the Firebase
   * web config (`apiKey`, `authDomain`, `projectId`, `appId`). Echoed
   * verbatim on public discovery. Omit for plain OIDC.
   */
  config?: Record<string, string>;
}

export interface IdpConfigPatch {
  enabled?: boolean;
  clientId?: string;
  allowedOrigins?: string[];
  comments?: string;
  /**
   * When set, REPLACES the stored provider config map wholesale (not
   * merged). Publishable values only — never secrets.
   */
  config?: Record<string, string>;
}

export class IdentityProviderConfigClient {
  constructor(
    private readonly http: HttpClient,
    private readonly realmId: string,
  ) {}

  /**
   * GET /identity-providers?platform_id={realmId}[&tenant_id=...].
   * Returns one page; a nil/absent `items` is normalized to an empty list
   * (matching `roles.list`).
   */
  async list(opts?: IdpConfigListOpts): Promise<IdpConfigListPage> {
    const raw = await this.http.request<unknown>({
      method: "GET",
      path: "/identity-providers",
      query: { platform_id: this.realmId, tenant_id: opts?.tenantId },
    });
    return normalizePage(raw);
  }

  /**
   * POST /identity-providers. `platform_id` is injected as the realm id.
   */
  async create(body: IdpConfigCreate): Promise<IdpConfig> {
    const wire: Record<string, unknown> = {
      platform_id: this.realmId,
      provider: body.provider,
      client_type: body.clientType,
      client_id: body.clientId,
    };
    if (body.tenantId !== undefined) wire["tenant_id"] = body.tenantId;
    if (body.allowedOrigins !== undefined) wire["allowed_origins"] = body.allowedOrigins;
    if (body.comments !== undefined) wire["comments"] = body.comments;
    if (body.config !== undefined) wire["config"] = body.config;
    return this.http.request<IdpConfig>({
      method: "POST",
      path: "/identity-providers",
      body: wire,
    });
  }

  /**
   * PATCH /identity-providers/{id}. Only set fields are sent; the server
   * returns `empty_patch` when nothing is provided.
   */
  async update(id: string, patch: IdpConfigPatch): Promise<IdpConfig> {
    const wire: Record<string, unknown> = {};
    if (patch.enabled !== undefined) wire["enabled"] = patch.enabled;
    if (patch.clientId !== undefined) wire["client_id"] = patch.clientId;
    if (patch.allowedOrigins !== undefined) wire["allowed_origins"] = patch.allowedOrigins;
    if (patch.comments !== undefined) wire["comments"] = patch.comments;
    if (patch.config !== undefined) wire["config"] = patch.config;
    return this.http.request<IdpConfig>({
      method: "PATCH",
      path: `/identity-providers/${encodeURIComponent(id)}`,
      body: wire,
    });
  }

  /** DELETE /identity-providers/{id}. Returns the deleted ack. */
  async delete(id: string): Promise<{ status: "deleted" }> {
    const out = await this.http.request<{ status?: string } | undefined>({
      method: "DELETE",
      path: `/identity-providers/${encodeURIComponent(id)}`,
    });
    return { status: (out?.status as "deleted") ?? "deleted" };
  }
}

function normalizePage(raw: unknown): IdpConfigListPage {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    return { items: [] };
  }
  const obj = raw as Record<string, unknown>;
  const items = Array.isArray(obj["items"]) ? (obj["items"] as IdpConfig[]) : [];
  return { items };
}
